File: backend/app/schemas/property.py
"""
Pydantic schemas for the property management API.

Covers: Property, Unit, Tenant, Lease, RentPayment, PropertyExpense,
MaintenanceRequest, SecurityDeposit, Mortgage, P&L, Metrics.
"""

from datetime import date, datetime
from typing import Optional, List
from pydantic import BaseModel, Field


# --- Property ---

class PropertyBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    address: Optional[str] = Field(None, max_length=500)
    property_type: str = Field(default="single_family")
    purchase_price: Optional[float] = None
    purchase_date: Optional[date] = None
    current_value: Optional[float] = None
    notes: Optional[str] = None
    is_active: bool = True


class PropertyCreate(PropertyBase):
    pass


class PropertyUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    address: Optional[str] = Field(None, max_length=500)
    property_type: Optional[str] = None
    purchase_price: Optional[float] = None
    purchase_date: Optional[date] = None
    current_value: Optional[float] = None
    notes: Optional[str] = None
    is_active: Optional[bool] = None


class PropertyResponse(PropertyBase):
    id: int
    total_monthly_rent: float
    unit_count: int
    occupied_unit_count: int
    vacancy_rate: float
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# --- Unit ---

class UnitBase(BaseModel):
    unit_number: str = Field(..., min_length=1, max_length=50)
    bedrooms: Optional[int] = Field(None, ge=0)
    bathrooms: Optional[float] = Field(None, ge=0)
    sqft: Optional[int] = Field(None, ge=0)
    monthly_rent: Optional[float] = Field(None, ge=0)


class UnitCreate(UnitBase):
    pass


class UnitUpdate(BaseModel):
    unit_number: Optional[str] = Field(None, min_length=1, max_length=50)
    bedrooms: Optional[int] = Field(None, ge=0)
    bathrooms: Optional[float] = Field(None, ge=0)
    sqft: Optional[int] = Field(None, ge=0)
    monthly_rent: Optional[float] = Field(None, ge=0)
    is_active: Optional[bool] = None


class UnitResponse(UnitBase):
    id: int
    property_id: int
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# --- Tenant ---

class TenantBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    email: Optional[str] = Field(None, max_length=200)
    phone: Optional[str] = Field(None, max_length=30)
    move_in_date: Optional[date] = None
    move_out_date: Optional[date] = None
    notes: Optional[str] = None
    is_active: bool = True


class TenantCreate(TenantBase):
    pass


class TenantUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    email: Optional[str] = Field(None, max_length=200)
    phone: Optional[str] = Field(None, max_length=30)
    move_in_date: Optional[date] = None
    move_out_date: Optional[date] = None
    notes: Optional[str] = None
    is_active: Optional[bool] = None


class TenantResponse(TenantBase):
    id: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# --- Lease ---

class LeaseBase(BaseModel):
    unit_id: int
    tenant_id: int
    start_date: date
    end_date: date
    monthly_rent: float = Field(..., gt=0)
    security_deposit: Optional[float] = Field(None, ge=0)
    terms_notes: Optional[str] = None
    status: str = Field(default="active")


class LeaseCreate(LeaseBase):
    pass


class LeaseUpdate(BaseModel):
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    monthly_rent: Optional[float] = Field(None, gt=0)
    security_deposit: Optional[float] = Field(None, ge=0)
    terms_notes: Optional[str] = None
    status: Optional[str] = None


class LeaseResponse(LeaseBase):
    id: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# --- Rent Payment ---

class RentPaymentBase(BaseModel):
    lease_id: int
    period_month: str = Field(..., pattern=r"^\d{4}-\d{2}$")
    amount_due: float = Field(..., gt=0)
    amount_paid: float = Field(default=0.0, ge=0)
    paid_date: Optional[date] = None
    status: str = Field(default="pending")
    late_fee: float = Field(default=0.0, ge=0)
    notes: Optional[str] = Field(None, max_length=500)


class RentPaymentCreate(RentPaymentBase):
    pass


class RentPaymentUpdate(BaseModel):
    amount_paid: Optional[float] = Field(None, ge=0)
    paid_date: Optional[date] = None
    status: Optional[str] = None
    late_fee: Optional[float] = Field(None, ge=0)
    notes: Optional[str] = Field(None, max_length=500)


class RentPaymentResponse(RentPaymentBase):
    id: int
    balance_due: float
    created_at: datetime

    model_config = {"from_attributes": True}


# --- Property Expense ---

class PropertyExpenseBase(BaseModel):
    property_id: int
    unit_id: Optional[int] = None
    category: str = Field(default="other")
    amount: float = Field(..., gt=0)
    date: date
    vendor: Optional[str] = Field(None, max_length=200)
    description: Optional[str] = Field(None, max_length=500)
    is_capex: bool = False


class PropertyExpenseCreate(PropertyExpenseBase):
    pass


class PropertyExpenseUpdate(BaseModel):
    category: Optional[str] = None
    amount: Optional[float] = Field(None, gt=0)
    date: Optional[date] = None
    vendor: Optional[str] = Field(None, max_length=200)
    description: Optional[str] = Field(None, max_length=500)
    is_capex: Optional[bool] = None


class PropertyExpenseResponse(PropertyExpenseBase):
    id: int
    created_at: datetime

    model_config = {"from_attributes": True}


# --- Maintenance Request ---

class MaintenanceRequestBase(BaseModel):
    property_id: int
    unit_id: int
    tenant_id: Optional[int] = None
    description: str = Field(..., min_length=1)
    priority: str = Field(default="medium")
    status: str = Field(default="open")
    created_date: date
    completed_date: Optional[date] = None
    vendor_name: Optional[str] = Field(None, max_length=200)
    cost: Optional[float] = Field(None, ge=0)


class MaintenanceRequestCreate(MaintenanceRequestBase):
    pass


class MaintenanceRequestUpdate(BaseModel):
    description: Optional[str] = Field(None, min_length=1)
    priority: Optional[str] = None
    status: Optional[str] = None
    completed_date: Optional[date] = None
    vendor_name: Optional[str] = Field(None, max_length=200)
    cost: Optional[float] = Field(None, ge=0)


class MaintenanceRequestResponse(MaintenanceRequestBase):
    id: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# --- Security Deposit ---

class SecurityDepositBase(BaseModel):
    lease_id: int
    amount: float = Field(..., gt=0)
    date_received: date
    interest_rate: float = Field(default=0.0, ge=0)
    deductions_json: Optional[str] = None
    refund_amount: Optional[float] = Field(None, ge=0)
    refund_date: Optional[date] = None


class SecurityDepositCreate(SecurityDepositBase):
    pass


class SecurityDepositUpdate(BaseModel):
    amount: Optional[float] = Field(None, gt=0)
    interest_rate: Optional[float] = Field(None, ge=0)
    deductions_json: Optional[str] = None
    refund_amount: Optional[float] = Field(None, ge=0)
    refund_date: Optional[date] = None


class SecurityDepositResponse(SecurityDepositBase):
    id: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# --- Mortgage ---

class MortgageBase(BaseModel):
    property_id: int
    lender: Optional[str] = Field(None, max_length=200)
    original_amount: float = Field(..., gt=0)
    current_balance: float = Field(..., ge=0)
    interest_rate: float = Field(..., ge=0)
    monthly_payment: float = Field(..., gt=0)
    start_date: Optional[date] = None
    term_years: Optional[int] = Field(None, gt=0)
    is_active: bool = True


class MortgageCreate(MortgageBase):
    pass


class MortgageUpdate(BaseModel):
    lender: Optional[str] = Field(None, max_length=200)
    current_balance: Optional[float] = Field(None, ge=0)
    interest_rate: Optional[float] = Field(None, ge=0)
    monthly_payment: Optional[float] = Field(None, gt=0)
    is_active: Optional[bool] = None


class MortgageResponse(MortgageBase):
    id: int
    ltv_ratio: float
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# --- Analytical Responses ---

class RentRollEntry(BaseModel):
    unit_id: int
    unit_number: str
    tenant_name: Optional[str] = None
    lease_id: Optional[int] = None
    monthly_rent: float
    status: str  # "occupied" | "vacant"
    lease_end: Optional[date] = None


class RentRollResponse(BaseModel):
    property_id: int
    property_name: str
    total_potential_rent: float
    total_collected: float
    entries: List[RentRollEntry]


class PropertyPNLResponse(BaseModel):
    property_id: int
    period_start: date
    period_end: date
    total_income: float
    total_expenses: float
    net_operating_income: float
    expense_breakdown: List["ExpenseBreakdownEntry"]


class ExpenseBreakdownEntry(BaseModel):
    category: str
    amount: float


class PropertyMetricsResponse(BaseModel):
    property_id: int
    noi: float  # Net Operating Income (annual)
    cash_flow: float  # NOI - mortgage payments (annual)
    cap_rate: Optional[float] = None  # NOI / current_value * 100
    cash_on_cash: Optional[float] = None  # cash_flow / total_cash_invested * 100
    ltv: Optional[float] = None  # mortgage_balance / current_value * 100
    dscr: Optional[float] = None  # NOI / annual_debt_service


class VacancyEntry(BaseModel):
    property_id: int
    property_name: str
    unit_id: int
    unit_number: str
    monthly_rent: float
    days_vacant: int
    lost_income: float


class VacancyResponse(BaseModel):
    total_vacant_units: int
    total_lost_income: float
    entries: List[VacancyEntry]


# --- Intelligence ---

class VacancyTrendResponse(BaseModel):
    property_id: int
    avg_vacancy_days: float = 0
    ewma_vacancy_days: float = 0
    trend: str  # increasing | decreasing | stable | insufficient_data
    sample_count: int = 0
    confidence: float = 0.0
    current_vacancy_rate: float = 0.0

class MaintenanceForecastResponse(BaseModel):
    property_id: int
    monthly_avg: float = 0
    ewma_monthly: float = 0
    current_month_spend: float = 0
    projected_month_spend: float = 0
    trend: str
    sample_count: int = 0
    confidence: float = 0.0

class PropertyInsightItem(BaseModel):
    type: str
    level: str  # info | warning | alert
    message: str
    reasoning: str

class CollectionHealthResponse(BaseModel):
    property_id: int
    on_time_rate: float = 0
    late_rate: float = 0
    total_payments: int = 0
    on_time_count: int = 0
    late_count: int = 0
    partial_count: int = 0
    confidence: float = 0.0

class PropertyIntelligenceResponse(BaseModel):
    property_id: int
    vacancy: VacancyTrendResponse
    maintenance: MaintenanceForecastResponse
    collection: CollectionHealthResponse
    insights: List[PropertyInsightItem]

class PortfolioScoreComponent(BaseModel):
    vacancy: float = 0
    collection: float = 0
    maintenance: float = 0
    noi: float = 0

class PortfolioScoreResponse(BaseModel):
    score: int = 0
    components: PortfolioScoreComponent
    property_count: int = 0
    avg_vacancy_rate: float = 0
    avg_collection_rate: float = 0
    confidence: float = 0.0

File: src/api/property.ts
/**
 * Property Management API — properties, units, tenants, leases,
 * rent payments, expenses, maintenance, deposits, mortgages, analytics.
 */

import { request } from './core';

// =============================================================================
// TYPES
// =============================================================================

export interface PropertyResponse {
  id: number;
  name: string;
  address: string | null;
  property_type: string;
  purchase_price: number | null;
  purchase_date: string | null;
  current_value: number | null;
  notes: string | null;
  is_active: boolean;
  total_monthly_rent: number;
  unit_count: number;
  occupied_unit_count: number;
  vacancy_rate: number;
  created_at: string;
  updated_at: string;
}

export interface PropertyCreate {
  name: string;
  address?: string;
  property_type?: string;
  purchase_price?: number;
  purchase_date?: string;
  current_value?: number;
  notes?: string;
}

export interface PropertyUpdate {
  name?: string;
  address?: string;
  property_type?: string;
  purchase_price?: number;
  purchase_date?: string;
  current_value?: number;
  notes?: string;
  is_active?: boolean;
}

export interface UnitResponse {
  id: number;
  property_id: number;
  unit_number: string;
  bedrooms: number | null;
  bathrooms: number | null;
  sqft: number | null;
  monthly_rent: number | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface UnitCreate {
  unit_number: string;
  bedrooms?: number;
  bathrooms?: number;
  sqft?: number;
  monthly_rent?: number;
}

export interface UnitUpdate {
  unit_number?: string;
  bedrooms?: number;
  bathrooms?: number;
  sqft?: number;
  monthly_rent?: number;
  is_active?: boolean;
}

export interface TenantResponse {
  id: number;
  name: string;
  email: string | null;
  phone: string | null;
  move_in_date: string | null;
  move_out_date: string | null;
  notes: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface TenantCreate {
  name: string;
  email?: string;
  phone?: string;
  move_in_date?: string;
  move_out_date?: string;
  notes?: string;
}

export interface TenantUpdate {
  name?: string;
  email?: string;
  phone?: string;
  move_in_date?: string;
  move_out_date?: string;
  notes?: string;
  is_active?: boolean;
}

export interface LeaseResponse {
  id: number;
  unit_id: number;
  tenant_id: number;
  start_date: string;
  end_date: string;
  monthly_rent: number;
  security_deposit: number | null;
  terms_notes: string | null;
  status: string;
  created_at: string;
  updated_at: string;
}

export interface LeaseCreate {
  unit_id: number;
  tenant_id: number;
  start_date: string;
  end_date: string;
  monthly_rent: number;
  security_deposit?: number;
  terms_notes?: string;
  status?: string;
}

export interface LeaseUpdate {
  start_date?: string;
  end_date?: string;
  monthly_rent?: number;
  security_deposit?: number;
  terms_notes?: string;
  status?: string;
}

export interface RentPaymentResponse {
  id: number;
  lease_id: number;
  period_month: string;
  amount_due: number;
  amount_paid: number;
  paid_date: string | null;
  status: string;
  late_fee: number;
  notes: string | null;
  balance_due: number;
  created_at: string;
}

export interface RentPaymentCreate {
  lease_id: number;
  period_month: string;
  amount_due: number;
  amount_paid?: number;
  paid_date?: string;
  status?: string;
  late_fee?: number;
  notes?: string;
}

export interface RentPaymentUpdate {
  amount_paid?: number;
  paid_date?: string;
  status?: string;
  late_fee?: number;
  notes?: string;
}

export interface PropertyExpenseResponse {
  id: number;
  property_id: number;
  unit_id: number | null;
  category: string;
  amount: number;
  date: string;
  vendor: string | null;
  description: string | null;
  is_capex: boolean;
  created_at: string;
}

export interface PropertyExpenseCreate {
  property_id: number;
  unit_id?: number;
  category?: string;
  amount: number;
  date: string;
  vendor?: string;
  description?: string;
  is_capex?: boolean;
}

export interface PropertyExpenseUpdate {
  category?: string;
  amount?: number;
  date?: string;
  vendor?: string;
  description?: string;
  is_capex?: boolean;
}

export interface MaintenanceRequestResponse {
  id: number;
  property_id: number;
  unit_id: number;
  tenant_id: number | null;
  description: string;
  priority: string;
  status: string;
  created_date: string;
  completed_date: string | null;
  vendor_name: string | null;
  cost: number | null;
  created_at: string;
  updated_at: string;
}

export interface MaintenanceRequestCreate {
  property_id: number;
  unit_id: number;
  tenant_id?: number;
  description: string;
  priority?: string;
  status?: string;
  created_date: string;
  completed_date?: string;
  vendor_name?: string;
  cost?: number;
}

export interface MaintenanceRequestUpdate {
  description?: string;
  priority?: string;
  status?: string;
  completed_date?: string;
  vendor_name?: string;
  cost?: number;
}

export interface SecurityDepositResponse {
  id: number;
  lease_id: number;
  amount: number;
  date_received: string;
  interest_rate: number;
  deductions_json: string | null;
  refund_amount: number | null;
  refund_date: string | null;
  created_at: string;
  updated_at: string;
}

export interface SecurityDepositCreate {
  lease_id: number;
  amount: number;
  date_received: string;
  interest_rate?: number;
  deductions_json?: string;
  refund_amount?: number;
  refund_date?: string;
}

export interface SecurityDepositUpdate {
  amount?: number;
  interest_rate?: number;
  deductions_json?: string;
  refund_amount?: number;
  refund_date?: string;
}

export interface MortgageResponse {
  id: number;
  property_id: number;
  lender: string | null;
  original_amount: number;
  current_balance: number;
  interest_rate: number;
  monthly_payment: number;
  start_date: string | null;
  term_years: number | null;
  is_active: boolean;
  ltv_ratio: number;
  created_at: string;
  updated_at: string;
}

export interface MortgageCreate {
  property_id: number;
  lender?: string;
  original_amount: number;
  current_balance: number;
  interest_rate: number;
  monthly_payment: number;
  start_date?: string;
  term_years?: number;
}

export interface MortgageUpdate {
  lender?: string;
  current_balance?: number;
  interest_rate?: number;
  monthly_payment?: number;
  is_active?: boolean;
}

// Analytics types
export interface RentRollEntry {
  unit_id: number;
  unit_number: string;
  tenant_name: string | null;
  lease_id: number | null;
  monthly_rent: number;
  status: string;
  lease_end: string | null;
}

export interface RentRollResponse {
  property_id: number;
  property_name: string;
  total_potential_rent: number;
  total_collected: number;
  entries: RentRollEntry[];
}

export interface ExpenseBreakdownEntry {
  category: string;
  amount: number;
}

export interface PropertyPNLResponse {
  property_id: number;
  period_start: string;
  period_end: string;
  total_income: number;
  total_expenses: number;
  net_operating_income: number;
  expense_breakdown: ExpenseBreakdownEntry[];
}

export interface PropertyMetricsResponse {
  property_id: number;
  noi: number;
  cash_flow: number;
  cap_rate: number | null;
  cash_on_cash: number | null;
  ltv: number | null;
  dscr: number | null;
}

export interface VacancyEntry {
  property_id: number;
  property_name: string;
  unit_id: number;
  unit_number: string;
  monthly_rent: number;
  days_vacant: number;
  lost_income: number;
}

export interface VacancyResponse {
  total_vacant_units: number;
  total_lost_income: number;
  entries: VacancyEntry[];
}

// Intelligence types
export interface VacancyTrendResponse {
  property_id: number;
  avg_vacancy_days: number;
  ewma_vacancy_days: number;
  trend: 'increasing' | 'decreasing' | 'stable' | 'insufficient_data';
  sample_count: number;
  confidence: number;
  current_vacancy_rate: number;
}

export interface MaintenanceForecastResponse {
  property_id: number;
  monthly_avg: number;
  ewma_monthly: number;
  current_month_spend: number;
  projected_month_spend: number;
  trend: 'increasing' | 'decreasing' | 'stable' | 'insufficient_data';
  sample_count: number;
  confidence: number;
}

export interface CollectionHealthResponse {
  property_id: number;
  on_time_rate: number;
  late_rate: number;
  total_payments: number;
  on_time_count: number;
  late_count: number;
  partial_count: number;
  confidence: number;
}

export interface PropertyInsightItem {
  type: string;
  level: 'info' | 'warning' | 'alert';
  message: string;
  reasoning: string;
}

export interface PropertyIntelligenceResponse {
  property_id: number;
  vacancy: VacancyTrendResponse;
  maintenance: MaintenanceForecastResponse;
  collection: CollectionHealthResponse;
  insights: PropertyInsightItem[];
}

export interface PortfolioScoreComponent {
  vacancy: number;
  collection: number;
  maintenance: number;
  noi: number;
}

export interface PortfolioScoreResponse {
  score: number;
  components: PortfolioScoreComponent;
  property_count: number;
  avg_vacancy_rate: number;
  avg_collection_rate: number;
  confidence: number;
}

// =============================================================================
// API NAMESPACE
// =============================================================================

export const propertyApi = {
  // Properties
  listProperties: (activeOnly = true) =>
    request<PropertyResponse[]>(`/property/properties?active_only=${activeOnly}`),
  createProperty: (data: PropertyCreate) =>
    request<PropertyResponse>('/property/properties', { method: 'POST', body: data }),
  getProperty: (id: number) =>
    request<PropertyResponse>(`/property/properties/${id}`),
  updateProperty: (id: number, data: PropertyUpdate) =>
    request<PropertyResponse>(`/property/properties/${id}`, { method: 'PUT', body: data }),
  deleteProperty: (id: number) =>
    request<void>(`/property/properties/${id}`, { method: 'DELETE' }),

  // Units
  listUnits: (propertyId: number) =>
    request<UnitResponse[]>(`/property/properties/${propertyId}/units`),
  createUnit: (propertyId: number, data: UnitCreate) =>
    request<UnitResponse>(`/property/properties/${propertyId}/units`, { method: 'POST', body: data }),
  updateUnit: (unitId: number, data: UnitUpdate) =>
    request<UnitResponse>(`/property/units/${unitId}`, { method: 'PUT', body: data }),
  deleteUnit: (unitId: number) =>
    request<void>(`/property/units/${unitId}`, { method: 'DELETE' }),

  // Tenants
  listTenants: (activeOnly = true) =>
    request<TenantResponse[]>(`/property/tenants?active_only=${activeOnly}`),
  createTenant: (data: TenantCreate) =>
    request<TenantResponse>('/property/tenants', { method: 'POST', body: data }),
  getTenant: (id: number) =>
    request<TenantResponse>(`/property/tenants/${id}`),
  updateTenant: (id: number, data: TenantUpdate) =>
    request<TenantResponse>(`/property/tenants/${id}`, { method: 'PUT', body: data }),
  deleteTenant: (id: number) =>
    request<void>(`/property/tenants/${id}`, { method: 'DELETE' }),

  // Leases
  listLeases: (status?: string) => {
    const params = status ? `?status=${status}` : '';
    return request<LeaseResponse[]>(`/property/leases${params}`);
  },
  createLease: (data: LeaseCreate) =>
    request<LeaseResponse>('/property/leases', { method: 'POST', body: data }),
  getLease: (id: number) =>
    request<LeaseResponse>(`/property/leases/${id}`),
  updateLease: (id: number, data: LeaseUpdate) =>
    request<LeaseResponse>(`/property/leases/${id}`, { method: 'PUT', body: data }),
  getExpiringLeases: (days = 90) =>
    request<LeaseResponse[]>(`/property/leases/expiring?days=${days}`),
  renewLease: (leaseId: number, data: LeaseCreate) =>
    request<LeaseResponse>(`/property/leases/${leaseId}/renew`, { method: 'POST', body: data }),

  // Rent payments
  listRentPayments: (leaseId?: number, status?: string) => {
    const params = new URLSearchParams();
    if (leaseId) params.append('lease_id', String(leaseId));
    if (status) params.append('status', status);
    const qs = params.toString();
    return request<RentPaymentResponse[]>(`/property/rent-payments${qs ? `?${qs}` : ''}`);
  },
  createRentPayment: (data: RentPaymentCreate) =>
    request<RentPaymentResponse>('/property/rent-payments', { method: 'POST', body: data }),
  updateRentPayment: (id: number, data: RentPaymentUpdate) =>
    request<RentPaymentResponse>(`/property/rent-payments/${id}`, { method: 'PUT', body: data }),
  getOverduePayments: () =>
    request<RentPaymentResponse[]>('/property/rent-payments/overdue'),

  // Expenses
  createExpense: (data: PropertyExpenseCreate) =>
    request<PropertyExpenseResponse>('/property/expenses', { method: 'POST', body: data }),
  listExpenses: (propertyId: number, category?: string) => {
    const params = category ? `?category=${category}` : '';
    return request<PropertyExpenseResponse[]>(`/property/properties/${propertyId}/expenses${params}`);
  },
  updateExpense: (id: number, data: PropertyExpenseUpdate) =>
    request<PropertyExpenseResponse>(`/property/expenses/${id}`, { method: 'PUT', body: data }),
  deleteExpense: (id: number) =>
    request<void>(`/property/expenses/${id}`, { method: 'DELETE' }),

  // Maintenance
  listMaintenance: (propertyId?: number, status?: string) => {
    const params = new URLSearchParams();
    if (propertyId) params.append('property_id', String(propertyId));
    if (status) params.append('status', status);
    const qs = params.toString();
    return request<MaintenanceRequestResponse[]>(`/property/maintenance${qs ? `?${qs}` : ''}`);
  },
  createMaintenance: (data: MaintenanceRequestCreate) =>
    request<MaintenanceRequestResponse>('/property/maintenance', { method: 'POST', body: data }),
  getMaintenance: (id: number) =>
    request<MaintenanceRequestResponse>(`/property/maintenance/${id}`),
  updateMaintenance: (id: number, data: MaintenanceRequestUpdate) =>
    request<MaintenanceRequestResponse>(`/property/maintenance/${id}`, { method: 'PUT', body: data }),
  getOpenMaintenance: () =>
    request<MaintenanceRequestResponse[]>('/property/maintenance/open'),

  // Security deposits
  listSecurityDeposits: (leaseId?: number) => {
    const params = leaseId ? `?lease_id=${leaseId}` : '';
    return request<SecurityDepositResponse[]>(`/property/security-deposits${params}`);
  },
  createSecurityDeposit: (data: SecurityDepositCreate) =>
    request<SecurityDepositResponse>('/property/security-deposits', { method: 'POST', body: data }),
  updateSecurityDeposit: (id: number, data: SecurityDepositUpdate) =>
    request<SecurityDepositResponse>(`/property/security-deposits/${id}`, { method: 'PUT', body: data }),

  // Mortgages
  listMortgages: (propertyId: number) =>
    request<MortgageResponse[]>(`/property/properties/${propertyId}/mortgages`),
  createMortgage: (data: MortgageCreate) =>
    request<MortgageResponse>('/property/mortgages', { method: 'POST', body: data }),
  updateMortgage: (id: number, data: MortgageUpdate) =>
    request<MortgageResponse>(`/property/mortgages/${id}`, { method: 'PUT', body: data }),

  // Analytics
  getRentRoll: (propertyId: number) =>
    request<RentRollResponse>(`/property/properties/${propertyId}/rent-roll`),
  getPNL: (propertyId: number, start: string, end: string) =>
    request<PropertyPNLResponse>(`/property/properties/${propertyId}/pnl?start=${start}&end=${end}`),
  getMetrics: (propertyId: number) =>
    request<PropertyMetricsResponse>(`/property/properties/${propertyId}/metrics`),
  getVacancies: () =>
    request<VacancyResponse>('/property/vacancies'),

  // Intelligence
  getPropertyIntelligence: (propertyId: number) =>
    request<PropertyIntelligenceResponse>(`/property/properties/${propertyId}/intelligence`),
  getVacancyTrend: (propertyId: number) =>
    request<VacancyTrendResponse>(`/property/properties/${propertyId}/vacancy-trend`),
  getMaintenanceForecast: (propertyId: number) =>
    request<MaintenanceForecastResponse>(`/property/properties/${propertyId}/maintenance-forecast`),
  getPortfolioScore: () =>
    request<PortfolioScoreResponse>('/property/portfolio-score'),
};
